### PR TITLE
Modify project structure in preparation for Upstream

### LINKED ISSUE #12: Submit the rust branch to upstream
We would like to submit the `rust-port` branch to upstream so that the Tock team would be able to offer feedback.
For this, change the following:

- [ ] delete the `python` folder
- [ ] move the `rust` folder to the root
- [ ] modify the workflow to build and test the rust version

### PR DESCRIPTION
**Pull request by eva-cosma**
This pull request solves issue #12 

This pull request contains the following changes:
- The python folder and its files have been removed
- The rust folder has been removed and its files have been moved to the root of the project
- A simple workflow has been added to build and test any push to "rust-port" or pull request with the target branch "rust-port"

To test the workflow, I've created a dummy pull request which includes a failing test.
https://github.com/george-cosma/tockloader/pull/2

### Questions
There are two issues I'd like to discuss regarding the workflow:
1. What should happen to the old workflows? Would it be appropriate to delete them?
2. Should the rust workflow be ran for any push/pull on any branch? Maybe all branches except main?

If my understanding of workflows is correct, github runs the workflows if the corresponding config files are found in the source branch of the pull request.

If this is true, then I think we could delete the python-specific workflows because a contributor to the rust version would pull from the "rust-port" branch, where the python workflows would not need to exist, and a contributor to the python version would pull from the "master" branch, where the rust workflow does not exist. Or so I think.

And in regards to my second question, I'm still uncertain of the advantages and disadvantages of testing all pull requests VS testing only the pull requests going towards the "rust-port" branch.

### CODE REVIEW

**Comment by alexandruradovici** [comment]
1. As we do not merge this into master, we can delete the workflows that are python related. As soon as this is merged into master, the python version will be deprecated.
2. You have specified which branch the workflow affects, my understanding is that the workflow will run only on this branch, it might not run on branches forked from this. I would suggest not to put a branch name, merge this and test if the workflow works on a branched forked and does not affect the master branch.

**Comment by eva-cosma** [comment]
I've removed all branch restrictions on that workflow and it works as it should:
1. A fork of the rust version will be checked by the rust workflow on pull request (https://github.com/george-cosma/tockloader/pull/3)
2. A fork of the main version will be checked by the python workflow on pull request (https://github.com/george-cosma/tockloader/pull/4)

There is one issue: PR on rust-port trigger the tests twice: once for being a pull request, and one for being a push.
image

A possible fix would be to limit the push event to only the "rust-port" branch, but Pull Requests will be checked regardless of target.

**Comment by alexandruradovici** [comment]
This is how Tock does all the testing, so that should be fine.

**Comment by eva-cosma** [comment]
Alright, I'll leave it as-is then

File: .github/workflows/rust.yml
Comment (alexandruradovici) [inline .github/workflows/rust.yml:1]: Add a `format` text that verifies the formatting of the source code. Take a look at how Tock does this.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: .github/workflows/rust.yml
Comment (eva-cosma) [reply .github/workflows/rust.yml:1]: Yep, I was also thinking that format should be checked, though looking through autoblack workflow for the python version of tockloader, it also automatically tries to add a new commit to the Pull Request with the formatted text. 

Should we also try to implement this, or leave it as a simple "fail if formatting fails"?

Comment (alexandruradovici) [reply .github/workflows/rust.yml:1]: No, just fail.

File: .github/workflows/rust.yml
Comment (alexandruradovici) [inline .github/workflows/rust.yml]: Add a newline

File: tools/run_clippy.sh
Comment (alexandruradovici) [inline tools/run_clippy.sh]: Add a newline

File: tools/run_fmt_check.sh
Comment (alexandruradovici) [inline tools/run_fmt_check.sh]: Add newline

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by eva-cosma** [comment]
I've looked at how Tock does their CI pipeline, and I must admit I'm quite a fan. 

They want to make sure all checks can be ran on the local version of the repository while keeping code repetition to a minimum. They do this by invoking rules in a Makefile.

In my latest commits, I've taken **heavy** inspiration from them. I've also made two more test forks to make sure they work properly.

Failing PR: https://github.com/george-cosma/tockloader/pull/5
Passing PR: https://github.com/george-cosma/tockloader/pull/6

Now, initially, I used a naming scheme similar to that of tock's Makefile, for the sake of consistency. Now, I'm not so certain, they might be a little verbose.

**Review by alexandruradovici — APPROVED** [review]

**Pull request merged**
